index.js

```diff
@@ -1,15 +1,57 @@
-const bcrypt = require('bcrypt');
+const crypto = require('crypto');
 const authMiddleware = require('express-basic-auth-safe');
 const debug = require('diagnostics')('authboot');
 
+const coalesce = (concurrency, action) => {
+  const trie = new Map();
+  function clean(key, value) {
+    const guesses = trie.get(key);
+    if (!guesses) return;
+    guesses.delete(value);
+    if (guesses.size === 0) {
+      trie.delete(key);
+    }
+  }
+  return (key, value, callback) => {
+    if (!trie.has(key)) {
+      trie.set(key, new Map());
+    }
+    const guesses = trie.get(key);
+    if (guesses.has(value)) {
+      guesses.get(value).push(callback);
+      return;
+    } else if (guesses.size > concurrency) {
+      debug(`brute force check failure for ${key}`);
```

Comment:
Would this mean that an attacker that knows our username could stop anyone using that username logging in by guessing if they can guess faster than we queue new legitimate requests?

Reply:
Probably still better than what was there before which just brought everything to its knees.

Reply:
yes this could lock a name out but it will be less strain on the server than what we had before. it does add to the point that we might want to make user names harder to guess though.